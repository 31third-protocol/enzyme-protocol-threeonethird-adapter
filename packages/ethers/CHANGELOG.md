# @enzymefinance/ethers

## 4.0.10

### Patch Changes

- 1f61b9af0: Release 4.0.10

## 4.0.9

### Patch Changes

- f82ac29f9: Latest protocol version

## 4.0.8

### Patch Changes

- c06baeab1: Add eth_sign for Gnosis Safe

## 4.0.7

### Patch Changes

- 8bb3e9ff6: Publish packages

## 4.0.5

### Patch Changes

- 8fe3d4a57: Publishing latest release

## 4.0.4

### Patch Changes

- 84c2e211e: Lock bn.js dependency version
- c030c92cd: Aave claimRewardsToSelf selector

## 4.0.3

### Patch Changes

- e8d978204: Upgrading ethers.js

## 4.0.2

### Patch Changes

- b4fcf8bc: Update packages after audit feedback and fixes.

## 4.0.1

### Patch Changes

- 2b54b39f: Add devtools to monorepo
