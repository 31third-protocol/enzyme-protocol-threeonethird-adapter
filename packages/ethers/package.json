{
  "name": "@enzymefinance/ethers",
  "version": "4.0.4",
  "description": "Technology Regulated and Operated Investment Funds",
  "author": "Enzyme Council DAO <council@enzyme.finance>",
  "license": "MIT",
  "homepage": "https://enzyme.finance",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/avantgardefinance/ethereum-devtools.git"
  },
  "bugs": {
    "url": "https://github.com/avantgardefinance/ethereum-devtools/issues"
  },
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    }
  },
  "scripts": {
    "build": "tsup",
    "clean": "rimraf .turbo .eslintcache node_modules dist",
    "lint": "eslint \"src/**/*.{js,ts}\" --cache",
    "format": "yarn lint --fix",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "ethers": "^5.6.8"
  },
  "devDependencies": {
    "@types/jest": "^27.4.0",
    "jest": "^27.4.7",
    "tsup": "^5.11.13",
    "typescript": "^4.5.5"
  },
  "tsup": {
    "entry": [
      "src/index.ts"
    ],
    "format": [
      "esm",
      "cjs"
    ],
    "dts": true,
    "splitting": true,
    "clean": true
  },
  "eslintConfig": {
    "extends": "@enzymefinance/eslint-config",
    "rules": {
      "@typescript-eslint/no-shadow": "off"
    },
    "parserOptions": {
      "project": [
        "./tsconfig.json"
      ]
    }
  }
}
