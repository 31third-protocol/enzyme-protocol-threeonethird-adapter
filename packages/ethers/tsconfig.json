{
  "extends": "../../tsconfig.json",
  "include": ["src"],
  "files": [],
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@enzymefinance/hardhat/types": ["../hardhat/src/types.ts"]
    }
  }
}
