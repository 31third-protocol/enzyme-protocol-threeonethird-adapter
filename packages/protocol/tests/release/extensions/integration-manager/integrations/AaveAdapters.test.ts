import { randomAddress } from '@enzymefinance/ethers';
import type { ComptrollerLib, ITestAddOnlyAddressListOwner, VaultLib } from '@enzymefinance/protocol';
import {
  aaveV2LendArgs,
  aaveV2RedeemArgs,
  ITestStandardToken,
  lendSelector,
  redeemSelector,
  SpendAssetsHandleType,
} from '@enzymefinance/protocol';
import type { ProtocolDeployment, SignerWithAddress } from '@enzymefinance/testutils';
import {
  aaveV2Lend,
  aaveV2Redeem,
  createNewFund,
  deployProtocolFixture,
  getAssetBalances,
  getAssetUnit,
  setAccountBalance,
} from '@enzymefinance/testutils';
import { BigNumber, utils } from 'ethers';
import type { IIntegrationAdapter } from 'src/codegen/IIntegrationAdapter';

const roundingBuffer = BigNumber.from(2);
let fork: ProtocolDeployment;
let aaveAdapter: IIntegrationAdapter, aTokenListOwner: ITestAddOnlyAddressListOwner;
let aToken: ITestStandardToken, underlying: ITestStandardToken;

const adapterKeys = ['aaveV2'];
describe.each(adapterKeys)('%s as adapter', (adapterKey) => {
  beforeEach(async () => {
    fork = await deployProtocolFixture();

    // Adapter-specific vars
    switch (adapterKey) {
      case 'aaveV2':
        aaveAdapter = fork.deployment.aaveV2Adapter;
        aTokenListOwner = fork.deployment.aaveV2ATokenListOwner;

        aToken = new ITestStandardToken(fork.config.aaveV2.atokens.ausdc, provider);
        underlying = new ITestStandardToken(fork.config.primitives.usdc, provider);

        break;
    }
  });

  describe('parseAssetsForAction', () => {
    it('does not allow a bad selector', async () => {
      const amount = utils.parseUnits('1', await underlying.decimals());

      const args = aaveV2LendArgs({
        aToken,
        amount,
      });

      await expect(
        aaveAdapter.parseAssetsForAction(randomAddress(), utils.randomBytes(4), args),
      ).rejects.toBeRevertedWith('_selector invalid');

      await expect(aaveAdapter.parseAssetsForAction(randomAddress(), lendSelector, args)).resolves.toBeTruthy();
    });

    it('generates expected output for lending', async () => {
      const amount = utils.parseUnits('1', await underlying.decimals());

      const args = aaveV2LendArgs({
        aToken,
        amount,
      });

      await expect(
        aaveAdapter.parseAssetsForAction(randomAddress(), utils.randomBytes(4), args),
      ).rejects.toBeRevertedWith('_selector invalid');

      const result = await aaveAdapter.parseAssetsForAction(randomAddress(), lendSelector, args);

      expect(result).toMatchFunctionOutput(aaveAdapter.parseAssetsForAction, {
        incomingAssets_: [aToken],
        minIncomingAssetAmounts_: [amount.sub(roundingBuffer)],
        spendAssetAmounts_: [amount],
        spendAssetsHandleType_: SpendAssetsHandleType.Transfer,
        spendAssets_: [underlying],
      });
    });

    it('generates expected output for redeeming', async () => {
      const amount = utils.parseUnits('1', await aToken.decimals());

      const args = aaveV2RedeemArgs({
        aToken,
        amount,
      });

      const result = await aaveAdapter.parseAssetsForAction(randomAddress(), redeemSelector, args);

      expect(result).toMatchFunctionOutput(aaveAdapter.parseAssetsForAction, {
        incomingAssets_: [underlying],
        minIncomingAssetAmounts_: [amount.sub(roundingBuffer)],
        spendAssetAmounts_: [amount],
        spendAssetsHandleType_: SpendAssetsHandleType.Transfer,
        spendAssets_: [aToken],
      });
    });
  });

  describe('actions', () => {
    let comptrollerProxy: ComptrollerLib, vaultProxy: VaultLib;
    let fundOwner: SignerWithAddress;

    beforeEach(async () => {
      [fundOwner] = fork.accounts;

      const newFundRes = await createNewFund({
        denominationAsset: new ITestStandardToken(fork.config.primitives.usdc, fundOwner),
        fundDeployer: fork.deployment.fundDeployer,
        fundOwner,
        signer: fundOwner,
      });

      comptrollerProxy = newFundRes.comptrollerProxy;
      vaultProxy = newFundRes.vaultProxy;

      // Seed the vault with the underlying asset
      const seedAmount = (await getAssetUnit(underlying)).mul(10);
      await setAccountBalance({ account: vaultProxy, amount: seedAmount, provider, token: underlying });
    });

    describe('lend', () => {
      let amount: BigNumber;

      beforeEach(async () => {
        amount = (await underlying.balanceOf(vaultProxy)).div(3);
      });

      it('happy path: unregistered aToken', async () => {
        const [preTxIncomingAssetBalance, preTxOutgoingAssetBalance] = await getAssetBalances({
          account: vaultProxy,
          assets: [aToken, underlying],
        });

        const receipt = await aaveV2Lend({
          aToken,
          aaveV2Adapter: aaveAdapter,
          amount,
          comptrollerProxy,
          fundOwner,
          integrationManager: fork.deployment.integrationManager,
        });

        const [postTxIncomingAssetBalance, postTxOutgoingAssetBalance] = await getAssetBalances({
          account: vaultProxy,
          assets: [aToken, underlying],
        });

        expect(postTxIncomingAssetBalance).toBeAroundBigNumber(preTxIncomingAssetBalance.add(amount), roundingBuffer);
        expect(postTxOutgoingAssetBalance).toBeAroundBigNumber(preTxOutgoingAssetBalance.sub(amount), roundingBuffer);

        expect(receipt).toMatchGasSnapshot(adapterKey);
      });

      it('happy path: registered aToken', async () => {
        // Register the aToken on the relevant list
        await aTokenListOwner.addValidatedItemsToList([aToken]);

        const receipt = await aaveV2Lend({
          aToken,
          aaveV2Adapter: aaveAdapter,
          amount,
          comptrollerProxy,
          fundOwner,
          integrationManager: fork.deployment.integrationManager,
        });

        expect(receipt).toMatchGasSnapshot(adapterKey);
      });
    });

    describe('redeem', () => {
      let amount: BigNumber;

      beforeEach(async () => {
        // Seed the vault with the aToken
        const seedAmount = (await getAssetUnit(aToken)).mul(10);
        await setAccountBalance({ account: vaultProxy, amount: seedAmount, provider, token: aToken });

        amount = seedAmount.div(3);
      });

      it('happy path: unregistered aToken', async () => {
        const [preTxIncomingAssetBalance, preTxOutgoingAssetBalance] = await getAssetBalances({
          account: vaultProxy,
          assets: [underlying, aToken],
        });

        const receipt = await aaveV2Redeem({
          aToken,
          aaveV2Adapter: aaveAdapter,
          amount,
          comptrollerProxy,
          fundOwner,
          integrationManager: fork.deployment.integrationManager,
        });

        const [postTxIncomingAssetBalance, postTxOutgoingAssetBalance] = await getAssetBalances({
          account: vaultProxy,
          assets: [underlying, aToken],
        });

        expect(postTxIncomingAssetBalance).toBeAroundBigNumber(preTxIncomingAssetBalance.add(amount), roundingBuffer);
        expect(postTxOutgoingAssetBalance).toBeAroundBigNumber(preTxOutgoingAssetBalance.sub(amount), roundingBuffer);

        // This can vary substantially for whatever reason
        expect(receipt).toMatchGasSnapshot(adapterKey);
      });

      it('happy path: registered aToken', async () => {
        // Register the aToken on the relevant list
        await aTokenListOwner.addValidatedItemsToList([aToken]);

        const receipt = await aaveV2Redeem({
          aToken,
          aaveV2Adapter: aaveAdapter,
          amount,
          comptrollerProxy,
          fundOwner,
          integrationManager: fork.deployment.integrationManager,
        });

        // This can vary substantially for whatever reason
        expect(receipt).toMatchGasSnapshot(adapterKey);
      });
    });
  });
});
