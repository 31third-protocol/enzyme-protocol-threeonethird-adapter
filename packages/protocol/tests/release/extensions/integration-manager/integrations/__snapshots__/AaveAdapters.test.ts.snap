// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aaveV2 as adapter actions lend happy path: registered aToken: aaveV2 1`] = `489011`;

exports[`aaveV2 as adapter actions lend happy path: unregistered aToken: aaveV2 1`] = `532448`;

exports[`aaveV2 as adapter actions redeem happy path: registered aToken: aaveV2 1`] = `446648`;

exports[`aaveV2 as adapter actions redeem happy path: unregistered aToken: aaveV2 1`] = `490085`;
